tls/extensions/s2n_client_psk.c

```diff
@@ -0,0 +1,250 @@
+        GUARD_AS_RESULT(s2n_blob_init(&identity, identity_data, identity_size));
+
+        /* TODO: Validate obfuscated_ticket_age when using session tickets.
+         *
+         * For identities established externally, an obfuscated_ticket_age of 0 SHOULD be
+         * used, and servers MUST ignore the value.
+         */
+        uint32_t obfuscated_ticket_age;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
+
+        /* TODO: Implement the callback to choose a PSK.
+         *
+         * When we don't have a callback configured to choose a PSK, we should fall back to accepting
+         * the first PSK identity that also exists in our list of supported PSKs. */
+        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &identity, &conn->psk_params.chosen_psk));
+
+        if (conn->psk_params.chosen_psk) {
+            conn->psk_params.chosen_psk_wire_index = wire_index;
+            return S2N_RESULT_OK;
+        }
+
+        wire_index++;
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_client_psk_recv_binder_list(struct s2n_connection *conn, struct s2n_blob *partial_client_hello,
+        struct s2n_stuffer *wire_binders_in)
+{
+    ENSURE_REF(conn);
+
+    uint8_t wire_index = 0;
+    while(s2n_stuffer_data_available(wire_binders_in)) {
+        uint8_t wire_binder_size;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint8(wire_binders_in, &wire_binder_size));
+
+        uint8_t *wire_binder_data;
+        ENSURE_REF(wire_binder_data = s2n_stuffer_raw_read(wire_binders_in, wire_binder_size));
+
+        struct s2n_blob wire_binder;
+        GUARD_AS_RESULT(s2n_blob_init(&wire_binder, wire_binder_data, wire_binder_size));
+
+        if (wire_index == conn->psk_params.chosen_psk_wire_index) {
+            GUARD_AS_RESULT(s2n_psk_verify_binder(conn, conn->psk_params.chosen_psk,
+                    partial_client_hello, &wire_binder));
+            return S2N_RESULT_OK;
+        }
+        wire_index++;
+    }
+    BAIL(S2N_ERR_BAD_MESSAGE);
+}
+
+static S2N_RESULT s2n_client_psk_recv_identities(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    ENSURE_REF(conn);
+
+    uint16_t identity_list_size;
+    GUARD_AS_RESULT(s2n_stuffer_read_uint16(extension, &identity_list_size));
+
+    uint8_t *identity_list_data;
+    ENSURE_REF(identity_list_data = s2n_stuffer_raw_read(extension, identity_list_size));
+
+    struct s2n_blob identity_list_blob;
+    GUARD_AS_RESULT(s2n_blob_init(&identity_list_blob, identity_list_data, identity_list_size));
+
+    struct s2n_stuffer identity_list;
+    GUARD_AS_RESULT(s2n_stuffer_init(&identity_list, &identity_list_blob));
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&identity_list, identity_list_blob.size));
+
+    return s2n_client_psk_recv_identity_list(conn, &identity_list);
+}
+
+static S2N_RESULT s2n_client_psk_recv_binders(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
```

Comment:
You must add
```c
ENSURE_REF(conn);
```
Otherwise, we might have a null-pointer dereference in line 217.